Add git branching strategy ADR fro review.

peteches opened the pull request:
Following the ways of working meeting yesterday I am submitting this git branch strategy for further discussion.

Commit:
Add git branching strategy ADR fro review.

ewastempel commented on doc/adr/0004-git-branching-strategy.md:
```diff
@@ -0,0 +1,84 @@
+
+  each _**category**_ describes the type of work to be found in that branch.
+
+  - _**feature**_ branches introduce some new functionality to the repository.
```
Is this what we decided in the meeting? My understanding was, that as long as we use JIRA ticket number, it's all fine. Meaning, we do not have to specify the type of work we are doing

peteches replied:
Ahh Ok, My line was pretty bad, So if I just remove the _**category**_ field it would be in line with what we agreed?

ewastempel replied:
I'd say, leave the category, but state, that it's optional and highlight that JIRA reference is compulsory.

peteches replied:
I have updated the wording to make clear that category is optional.

Would you mind reviewing the wording please @ewastempel?

ewastempel replied:
Looks good Pete!

Commit:
Make category optional.

As @ewastempel mentions in #4 category should be optional.

ewastempel reviewed (approved):
Looks good

spikeheap commented on doc/adr/0004-git-branching-strategy.md:
```diff
@@ -0,0 +1,86 @@
+  - personal/PMCCABE-1-experiment-with-vmextension-script-deployments
+
+Each branch should be created from `origin/master` then worked on and tested.
+When ready a Pull Request should be generated to merge back into `origin/master`
```
:+1: too _should_. We'll rarely want to branch of branches, but I have found that useful in the past to provide stepped reviews when there's not been a reviewer for a day or so rather than one big wall of code.

spikeheap commented on doc/adr/0004-git-branching-strategy.md:
```diff
@@ -0,0 +1,86 @@
+developed code. There will be a 1 to 1 relationship between deployed
+environments and repository brnaches. The only exception to this 1 to 1
+relationship will be the _**master**_ branch. The _**master**_ branch may be used to
+back any environment that needs to be live like.
```
Could we move this block below the branching strategy. It feels like the deployment of environments is secondary to the approach, and is more a consequence of having a consistent approach (i.e. we could adopt this strategy and take a completely different approach to managing the dev subscription, though I'm not suggesting we do)

spikeheap commented on doc/adr/0004-git-branching-strategy.md:
```diff
@@ -0,0 +1,86 @@
+
+  `<category>/<default-environment-reference>-<summary>`
+  OR
+  `<default-environment-reference>-<summary>`
```
I think we should use `jira-id` rather than `default-environment-reference` here, and add another variant for personal branches, in order to separate out the way we manage environments from our branching strategy a little. I think it also makes this easier to use as a reference later on.

For a different discussion, we'll probably want to deploy environments based on the full branch name, rather than a portion, because we could see multiple branches for the same issue. Decoupling the deployment strategy from branching strategy, but making sure they're compatible feels cleaner.

peteches replied:
jira-id makes sense, I was trying to decouple us from specific tools. But we can always update this later if we switch.

deploying off the full branch name will raise a whole heap of other issues. for instance to properly namespace resources in azure the name attribute will need to have the env-name prefixed onto it. we will quickly run into issues, some resources don't like certain characters, some only want lower case letters. There is almost certainly a character limit which, `infra-32-creating-a-new-thing` will seriously eat into. (that short name is 29 characters.) VM names for example have a limit of 64 characters.

spikeheap replied:
:+1: good point about the tooling. Maybe `issue-id` to keep it generic?

Sounds like there's a good discussion to be had around (auto)deployment models later on. Exciting times 🥳

peteches replied:
I like `issue-id`

Commit:
Switching default-environment-reference to jira-id

Implement suggestion by @spikeheap switching to use jira-id to be
clearer about what we expect.

Commit:
Move branch to env relationship to consequence.

@spikeheap has pointed out that the relationship between branches and
deployed environments is a consequence of having a solid branching
strategy not the actual decision.

spikeheap reviewed (approved):
:+1:, really liking the discussion that's gone on here

Commit:
decouple from specific tooling.

we may choose not to use jira in future so using issue-id instead keeps
it generic without linking us to a deployment strategy.
